### PR TITLE
Make drives a bit more customizable (for modders)

### PR DESCRIPTION
**Pull request by Rakete1111**
This PR makes the drives (hyper, scam and jump) more customizable for modders. For instance, you can now change the amount of fuel a hyperdrive uses per jump, make a drive that is both scram and jump, and stuff like that. Not much more really. It also fixes https://github.com/endless-sky/endless-sky/issues/1238. 

My code editor strips lines with just whitespaces, so I hope that's not a big problem. Also, I noticed several typos in the comments related to the drives, so I fixed them. If I misunderstood a comment, please tell me.
As for as I know, my code follows the style guide, but a lot of stuff (like scoped enum or unscoped enum?) are missing, so I used the convention already in the code base.

Any comments appreciated! Thanks

### CODE REVIEW

File: data/outfits.txt [hunk 1]
```diff
@@ -176,8 +177,9 @@ outfit "Scram Drive"
 	thumbnail "outfit/hyperdrive"
 	"mass" 30
 	"outfit space" -30
-	"scram drive" .2
-	"hyperdrive" 1
+	"scram drive" 1
+	deviation .2
```
Comment (EndrosG) [inline data/outfits.txt]: may be "scram deviation", "jump deviation" or something "longer"? -> Because the other attribute is "jump speed" and not just "speed".

File: source/DistanceMap.cpp [hunk 2]
```diff
@@ -41,7 +41,7 @@ DistanceMap::DistanceMap(const PlayerInfo &player, const System *center)
 {
 	if(!player.Flagship())
 		return;
-	
+
```
Comment (EndrosG) [inline source/DistanceMap.cpp]: Your editor removed all the tabs in "empty" lines.

File: source/Fleet.cpp [hunk 3]
```diff
@@ -96,30 +96,35 @@ void Fleet::Enter(const System &system, list<shared_ptr<Ship>> &ships, const Pla
...
 	// Where this fleet can come from depends on whether it is friendly to any
 	// planets in this system and whether it has jump drives.
 	vector<const System *> linkVector;
 	// Find out what the "best" jump method the fleet has is. Assume that if the
 	// others don't have that jump method, they are being carried as fighters.
 	// That is, content creators should avoid creating fleets with a mix of jump
 	// drives and hyperdrives.
-	int jumpType = 0;
+	bool hasJumpDrive = false;
+	double jumpType = 0.;
```
Comment (EndrosG) [inline source/Fleet.cpp]: for consistency, I would ask you to go ahead and change (rename) this then. jumpType is then no longer a "Type" but a "Cost" variable.
In Ship.h you defined a new enum JumpType for that purpose.
Now you have two pretty different types of "jumpType".
In this case, you only need the ability and do not need the cost at all. The entire Fleet.cpp contained only 5x the word "jumpType". So, if you change it to hasJumpDrive, you dont neven need your double jumpType (with the meaning of jumpCost) anymore.

Also, you could explicitely overcome the original comment in the lines above and provide a solution which handles "mixed" fleets.
I thought about doing that but will not do it right now in order to avoid code conflicts.
My approach would be to take `jumpType = min(jumpType, ship->GetJumpType())` for all ships which can't be carried. Otherwise the fighters and drones, which have jumpType=0 (or NONE), would mess up because of the min-operator.

File: source/DistanceMap.cpp (hunk 2 above)
Comment (Rakete1111) [inline source/DistanceMap.cpp]: @EndrosG I know, next time, I will disable that setting.

File: data/outfits.txt (hunk 1 above)
Comment (Rakete1111) [inline data/outfits.txt]: @EndrosG I went with "scram deviation" as it only deviates for the Scram Drive. Thanks :)

File: source/Fleet.cpp (hunk 3 above)
Comment (Rakete1111) [inline source/Fleet.cpp]: @EndrosG You're right, I completely missed the fact that I don't even need to get the cost :) I also made it so fleets can have multiple drives, but use another approach, which I think would also work without any additional code to handle fighters/drones.

File: source/Fleet.cpp [hunk 4]
```diff
@@ -107,24 +107,16 @@ void Fleet::Enter(const System &system, list<shared_ptr<Ship>> &ships, const Pla
...
-	if(jumpType)
+		jumpType |= ship->HyperspaceType();
+
+	if(jumpType != Ship::NONE)
 	{
 		// Don't try to make a fleet "enter" from another system if none of the
 		// ships have jump drives.
 		bool isWelcomeHere = !system.GetGovernment()->IsEnemy(government);
-		for(const System *neighbor : (hasJumpDrive ? system.Neighbors() : system.Links()))
+		for(const System *neighbor : system.Links())
```
Comment (EndrosG) [inline source/Fleet.cpp]: You removed the differentiation between fleets with Jump Drives and fleets without Jump Drives. :-1: 
The old condition allowed fleets with Jump Drives to jump in from any neighbor system, while other fleets had to you hyperlinks.

File: source/DistanceMap.cpp (hunk 2 above)
Comment (EndrosG) [inline source/DistanceMap.cpp]: Please dont get me wrong, but your pull will - as far as I know - never be merged if you dont fix that in this pull too. It is good to change it for the future, but if the author would merge this pull, all your trimming would be applied to the master branch.
I did not even look at 2 or 3 files of your initial commit because the diff was not loaded initially (because of that many changes).
So please fix this for your existing pull too.

File: source/Fleet.cpp (hunk 4 above)
Comment (Rakete1111) [inline source/Fleet.cpp]: @EndrosG I thought that it wasn't necessary, because you can't see where the ship with jump drive can from, so it doesn't matter. But this fails for one case: if the system has no hyperspace links but the fleet is equipped with only jump drives. Thanks for pointing that out! :)

File: source/DistanceMap.cpp (hunk 2 above)
Comment (Rakete1111) [inline source/DistanceMap.cpp]: @EndrosG If I make another commit which fixes the spaces, then it will also be huge. Should I close this and make a new clean PR, or should I do it anyways?

**Comment by EndrosG** [comment]
> @EndrosG If I make another commit which fixes the spaces, then it will also be huge. Should I close this and make a new clean PR, or should I do it anyways?

It is up to you. But if I were you, I would probably close this one and make a new one.

------

I still have difficulties overseeing all the changes, because of the many empty lines, which make it very hard to read and to get a compact view for the whole thing.

In particular, I am not even sure if the code fragments I've read will really work in every circumstance.

Thought one: You define JumpType as a bitwise flag operator.
In theory this means, that even a single outfit can provide various drive capabilities. At least the "capabilities" of a ship can reach at most 0x111 = 7 = JUMP | SCRAM | HYPER.
I am not sure if all of your checks take care of this, or if you eventually have some lines remaining where you e.g. would check if 0x111 == 0x100 == JUMP

Having another additional function, HyperspaceTypes(), which produces that 0x111 value, seems making the thing more complex. I already now have the "problem", that this part of the code uses more functions with very fixed assumptions than I would like to see.
Your approach initially sounds nice, but so far I have seen at least three major flaws:
1. It will break any existing mods. This change is incompatible with existing data files and while we may have control over the data files provided *with* the game (i.e. we can change the data files being distributed with the game version), we have no influence on existing mods.
Example: [Multi Drive in Shields Up plugin, found @Amazinite repo's](https://github.com/Amazinite/Shields-Up/blob/cf0ca2378bd045129f7afa341cda79d0211560e7/000-Shields%20Up/data/sentars.txt#L84) --- First, you modify the meaning of the existing `scram drive` tag. Second, I have not seen that your code provides default values. May be I oversaw that in the many lines :-/

2. You aim to give modders flexibility but dont implement it to the last.
If I were to create a unique jumpdrive which uses 90 fuel instead of 200, the properly modified game should be aware of that and prefer the jump drive capability over the hyperdrive capability then. (This is the reason why I dont like to have seventeen functions for one jump...)

3. You did not properly adjust the DistanceMap stuff. If I were to create a human experimental jump drive which would use let's say 800 fuel instead of 200 but would at least offer the jump capability, the pathfinding would not be aware of these super high cost.
```
void DistanceMap::Init(const System *center, const Ship *ship)
	// Find the route with lowest fuel use. If multiple routes use the same fuel,
	// choose the one with the fewest jumps (i.e. using jump drive rather than
	// hyperdrive). If multiple routes have the same fuel and the same number of
```
You did not take into account the particular values specified by any modder.

**Comment by Tadrix** [comment]
I suggest closing this and making new one. Remember, M.Z. has to check your PR. The less there is to check, the more chances your PR gets accepted.

**Comment by Rakete1111** [comment]
Thanks for all the suggestions! I'll make a new one then.

**Comment by jafdy** [comment]
@Tadrix This is a tiny one at only 3 commits. Having 2 PRs on the same thing is more annoying then having one with a small amount of comments and commits.
